packages/dd-trace/src/plugins/plugin.js

```diff
@@ -0,0 +1,52 @@
+'use strict'
+
+const dc = require('diagnostics_channel')
+const { storage } = require('../../../datadog-core')
+
+class Subscription {
+  constructor (event, handler) {
+    this._channel = dc.channel(event)
+    this._handler = handler
+  }
+
+  enable () {
+    this._channel.subscribe(this._handler)
+  }
+
+  disable () {
+    this._channel.unsubscribe(this._handler)
+  }
+}
+
+module.exports = class Plugin {
+  constructor () {
+    this._subscriptions = []
+    this._enabled = false
+    this._storeStack = []
+  }
+
+  enter (span, store) {
+    store = store || storage.getStore()
+    this._storeStack.push(store)
+    storage.enterWith({ ...store, span })
+  }
+
+  exit () {
+    storage.enterWith(this._storeStack.pop())
+  }
+
+  addSub (channelName, handler) {
+    this._subscriptions.push(new Subscription(channelName, handler))
+  }
+
+  configure (config) {
+    this.config = config
+    if (config.enabled && !this._enabled) {
```

Comment:
Let's move these 2 blocks into individual `enable` and `disable` methods and call the appropriate one based on the config.

Reply:
Enabling and disabling has to go through the config to be consistent.